XenAdmin/Dialogs/EnablePvsReadCachingDialog.cs

```diff
@@ -0,0 +1,184 @@
+            }
+        }
+
+        private void enableButton_Click(object sender, EventArgs e)
+        {
+            var siteItemSelected = (ToStringWrapper<PVS_site>) pvsSiteList.SelectedItem;
+            var siteSelected = siteItemSelected.item;
+
+            var actions = new List<AsyncAction>();
+
+            foreach (var vm in vms)
+            {
+                var action = GetAsyncActionForVm(vm, siteSelected);
+                if (action != null)
+                {
+                    actions.Add(action);
+                }
+            }
+
+            if (actions.Any())
+            {
+                var session = vms[0].Connection.Session;
+
+                if (actions.Count == 1)
+                {
+                    actions[0].RunExternal(session);
+                }
+                else
+                {
+                    new ParallelAction(
+                        Messages.ACTION_ENABLE_PVS_READ_CACHING,
+                        Messages.ACTION_ENABLING_PVS_READ_CACHING,
+                        Messages.ACTION_ENABLED_PVS_READ_CACHING,
+                        actions).RunExternal(session);
+                }
+            }
+        }
+
+        /// <summary>
+        /// If the VM can have PVS read-caching enabled, returns an action to do so. Else returns null
+        /// </summary>
+        /// <param name="vm"></param>
+        /// <param name="siteSelected"></param>
+        /// <param name="prepopulate"></param>
+        /// <returns></returns>
+        private AsyncAction GetAsyncActionForVm(VM vm, PVS_site siteSelected)
+        {
+            if (PvsProxyAlreadyEnabled(vm))
+            {
+                return null;
+            }
+
+            var vif = GetVifForPvsProxy(vm);
+            if (vif == null)
+            {
+                return null; // No VIF with device = 0, so can't enable
+            }
+
+            return new PvsProxyCreateAction(vm, siteSelected, vif);
+        }
+
+        /// <summary>
+        /// A VM can be enabled if there isn't already a PVS_Proxy on its VIF
+        /// </summary>
+        /// <param name="vm"></param>
+        /// <returns></returns>
+        private bool PvsProxyAlreadyEnabled(VM vm)
+        {
+            var connection = vm.Connection;
+
+            var pvsProxies = connection.Cache.PVS_proxies;
+
+            foreach (var pvsProxy in pvsProxies)
+            {
+                if(pvsProxy.VM.Equals(vm))
+                {
+                    return true; // Already got a PVS proxy on this vm
+                }
+            }
+
+            return false; // No PVS proxy is on this VM
+        }
+
+        /// <summary>
+        /// The VIF for a PVS Proxy is the one with VIF.device=0
+        /// </summary>
+        /// <param name="vm"></param>
+        private VIF GetVifForPvsProxy(VM vm)
+        {
+            var vifRefs = vm.VIFs;
+            var vifs = vm.Connection.ResolveAll(vifRefs);
+
+            foreach (var vif in vifs)

```

Comment:
This could be simplified into `return vifs.FirstOrDefault(vif => vif.device.Equals("0"));`